### PR TITLE
Update master from 2.8.7

### PR DESCRIPTION
**Pull request by AviAvni**

### CODE REVIEW

File: src/serializers/decoders/current/v11/decode_graph_schema.c [hunk 1]
```diff
@@ -30,15 +30,17 @@ static Schema *_RdbLoadSchema(RedisModuleIO *rdb, SchemaType type) {
...
 			uint stopwords_count = RedisModule_LoadUnsigned(rdb);
 			if(stopwords_count > 0) {
-				stopwords = array_new(char *, stopwords_count);
+				if(!already_loaded)
+					stopwords = array_new(char *, stopwords_count);
 				for (uint i = 0; i < stopwords_count; i++) {
 					char *stopword = RedisModule_LoadStringBuffer(rdb, NULL);
-					array_append(stopwords, rm_strdup(stopword));
+					if(!already_loaded)
+						array_append(stopwords, rm_strdup(stopword));
```
Comment (jeffreylovitz) [inline src/serializers/decoders/current/v11/decode_graph_schema.c:43]: I don't think this works; there's no place in which `stopwords` is retrieved in the `already_loaded` case - it is NULL.

**Review by jeffreylovitz — CHANGES_REQUESTED** [review]
This mostly looks great to me, one small comment area.

**Review by swilly22 — CHANGES_REQUESTED** [review]

File: tests/flow/test_index_scans.py [hunk 2]
```diff
@@ -527,3 +527,40 @@ def test18_index_scan_inside_apply(self):
...
+
+        result = redis_graph.query("MATCH (u:L1) WITH min(u.id) as id MATCH (u1:L1), (u2:L1) WHERE u1.id = 990000000262240069 AND (u2.id = 990000000262240070 OR u2.id = 990000000262240071) RETURN u1.id, u2.id")
+        expected_result = [[990000000262240069, 990000000262240070], [990000000262240069, 990000000262240071]]
+        self.env.assertEquals(result.result_set, expected_result)
+
+        result = redis_graph.query("MATCH (u:L1) WITH min(u.id) as id MATCH (u1:L1), (u2:L1) WHERE u1.id = id AND (u2.id = 990000000262240070 OR u2.id = 990000000262240071) RETURN u1.id, u2.id")
+        expected_result = [[990000000262240069, 990000000262240070], [990000000262240069, 990000000262240071]]
+        self.env.assertEquals(result.result_set, expected_result)
+
+        result = redis_graph.query("MATCH (u:L2) WHERE u.id1 = 990000000262240069 AND u.id2 = 990000000262240067 RETURN u.id1, u.id2")
```
[the remark this answers is not in the document]
Comment (AviAvni) [reply tests/flow/test_index_scans.py:570]: done

File: src/execution_plan/ops/op_node_by_index_scan.c [hunk 3]
```diff
@@ -172,7 +172,6 @@ static Record IndexScanConsumeFromChild(OpBase *opBase) {
 			// first call to consume, create query and iterator
```
[the remark this answers is not in the document]
Comment (AviAvni) [reply src/execution_plan/ops/op_node_by_index_scan.c:172]: done

File: src/serializers/decoders/prev/v10/decode_graph.c [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+			GrB_Index nvals;
+			RG_Matrix L = Graph_GetLabelMatrix(g, i);
+			RG_Matrix_nvals(&nvals, L);
+			GraphStatistics_IncNodeCount(&g->stats, i, nvals);
+		}
+
+		// make sure graph doesn't contains may pending changes
+		ASSERT(Graph_Pending(g) == false);
+
+		QueryCtx_Free(); // release thread-local variables
```
Comment (swilly22) [inline src/serializers/decoders/prev/v10/decode_graph.c:209]: now that `QueryCtx_SetGraphCtx` is commented out, see if we can remove the call to `GraphDecodeContext_Reset` ?

File: src/serializers/encoder/v11/encode_graph.c [hunk 5]
```diff
@@ -109,6 +111,7 @@ static PayloadInfo *_RdbSaveKeySchema
 	//  N * Payload info:
 	//      Encode state
 	//      Number of entities encoded in this state
+	//  Schema
```
Comment (swilly22) [inline src/serializers/encoder/v11/encode_graph.c:114]: Please relocate this comment (single line)

File: src/serializers/encoder/v11/encode_graph.c [hunk 6]
```diff
@@ -210,19 +213,18 @@ void RdbSaveGraph_v10
...
 		case ENCODE_STATE_EDGES:
-			RdbSaveEdges_v10(rdb, gc, payload.entities_count);
+			RdbSaveEdges_v11(rdb, gc, payload.entities_count);
 			break;
 		case ENCODE_STATE_DELETED_EDGES:
-			RdbSaveDeletedEdges_v10(rdb, gc, payload.entities_count);
+			RdbSaveDeletedEdges_v11(rdb, gc, payload.entities_count);
 			break;
 		case ENCODE_STATE_GRAPH_SCHEMA:
-			RdbSaveGraphSchema_v10(rdb, gc);
```
Comment (swilly22) [inline src/serializers/encoder/v11/encode_graph.c:225]: Please add a comment why is this case is empty.

**Review by swilly22 — CHANGES_REQUESTED** [review]

File: src/serializers/decoders/prev/v10/decode_graph.c (hunk 4 above)
Comment (AviAvni) [reply src/serializers/decoders/prev/v10/decode_graph.c:207]: we can't remove it it is set also in the header

Comment (swilly22) [reply src/serializers/decoders/prev/v10/decode_graph.c]: Call `QueryCtx_Free` at the end right before `return gc` apply to all decoders.

File: src/serializers/decoders/current/v11/decode_graph_schema.c [hunk 7]
```diff
@@ -30,15 +30,17 @@ static Schema *_RdbLoadSchema(RedisModuleIO *rdb, SchemaType type) {
 
 		if(index_type == IDX_FULLTEXT) {
 			char *lang = RedisModule_LoadStringBuffer(rdb, NULL);
-			language = rm_strdup(lang);
+			if(!already_loaded) language = rm_strdup(lang);
```
Comment (swilly22) [inline src/serializers/decoders/current/v11/decode_graph_schema.c:38]: Please reduce `if(!already_loaded)` as much as possible

File: src/serializers/decoders/current/v11/decode_graph_schema.c [hunk 8]
```diff
@@ -48,18 +50,25 @@ static Schema *_RdbLoadSchema(RedisModuleIO *rdb, SchemaType type) {
...
 			RedisModule_Free(field);
 		}
 
-		if(index_type == IDX_FULLTEXT) {
+		if(index_type == IDX_FULLTEXT && !already_loaded) {
 			idx->language = language;
 			idx->stopwords = stopwords;
 		}
 
 		index_count -= fields_count;
```
Comment (swilly22) [inline src/serializers/decoders/current/v11/decode_graph_schema.c:68]: Need to make a few changes to the way indexes are formatted, I think `index_count` should reflect the number of indexes and `fields_count` should reflect the number of fields within a given index.
`index_count -= fields_count;` is confusing.

File: src/serializers/decoders/current/v11/decode_graph_schema.c [hunk 9]
```diff
@@ -30,15 +30,17 @@ static Schema *_RdbLoadSchema(RedisModuleIO *rdb, SchemaType type) {
 
 		if(index_type == IDX_FULLTEXT) {
```
Comment (swilly22) [inline src/serializers/decoders/current/v11/decode_graph_schema.c:36]: Consider introducing 2 new functions:
1. decode FULLTEXT-INDEX
2.decode EXACTMATCH-INDEX

**Review by swilly22 — CHANGES_REQUESTED** [review]

File: src/index/index.c [hunk 10]
```diff
@@ -356,6 +356,30 @@ char **Index_GetStopwords
...
+// set indexed language
+void Index_SetLanguage
+(
+	Index *idx,
+	const char *language
+) {
+	ASSERT(idx != NULL);
+	ASSERT(language != NULL);
+
+	idx->language = rm_strdup(language);
```
Comment (swilly22) [inline src/index/index.c:368]: ASSERT(idx->language == NULL);

File: src/index/index.c [hunk 11]
```diff
@@ -356,6 +356,30 @@ char **Index_GetStopwords
...
+// set indexed stopwords
+void Index_SetStopwords
+(
+	Index *idx,
+	char **stopwords
+) {
+	ASSERT(idx != NULL);
+	ASSERT(stopwords != NULL);
+	
+	array_clone_with_cb(idx->stopwords, stopwords, rm_strdup);
```
Comment (swilly22) [inline src/index/index.c:380]: ASSERT(idx->stopwords == NULL);

File: src/serializers/decoders/current/v11/decode_graph_schema.c [hunk 12]
```diff
@@ -1,66 +1,111 @@
...
+	/* Format:
+	 * language
+	 * #stopwords - N
+	 * N * stopword
+	 * #properties - M
+	 * M * property */
+
+	char *language = NULL;
+	char **stopwords = NULL;
+
```
Comment (swilly22) [inline src/serializers/decoders/current/v11/decode_graph_schema.c:24]: remove empty line

File: src/serializers/decoders/current/v11/decode_graph_schema.c [hunk 13]
```diff
@@ -1,66 +1,111 @@
...
+	uint stopwords_count = RedisModule_LoadUnsigned(rdb);
+	if(stopwords_count > 0) {
+		stopwords = array_new(char *, stopwords_count);
+		for (uint i = 0; i < stopwords_count; i++) {
+			char *stopword = RedisModule_LoadStringBuffer(rdb, NULL);
+			array_append(stopwords, stopword);
+		}
+	}
+
+	Index *idx = NULL;
```
Comment (swilly22) [inline src/serializers/decoders/current/v11/decode_graph_schema.c:36]: move to line 21

File: src/serializers/decoders/current/v11/decode_graph_schema.c [hunk 14]
```diff
@@ -1,66 +1,111 @@
...
+	uint fields_count = RedisModule_LoadUnsigned(rdb);
+	for(uint i = 0; i < fields_count; i++) {
+		char *field = RedisModule_LoadStringBuffer(rdb, NULL);
+		if(!already_loaded) Schema_AddIndex(&idx, s, field, IDX_FULLTEXT);
+		RedisModule_Free(field);
+	}
+
+	if(!already_loaded) {
+		Index_SetLanguage(idx, lang);
```
Comment (swilly22) [inline src/serializers/decoders/current/v11/decode_graph_schema.c:45]: `ASSERT(idx != NULL);`

**Review by swilly22 — APPROVED** [review]
@AviAvni approved, please see additional comments

**Review by swilly22 — APPROVED** [review]

**Review by swilly22 — APPROVED** [review]

**Pull request merged**